Fix the summary of converted object types to be treated as HTML

Menrath opened the pull request:
Tries to solve https://github.com/mastodon/mastodon/issues/28455

This is me the first time writing Ruby code. Also I didn't have the time to test this and setup a proper development environment. Still I wanted to try to contribute on solving this issue.

Also I am not sure where Mastodon sanitized the HTML for the content (status.text) for non-converted object types. This should also be done here.

Menrath committed:
Converted object types summary (spoiler text) is treated as HTML.

Solves https://github.com/mastodon/mastodon/issues/28455

ClearlyClaire commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -372,7 +372,8 @@ def in_reply_to_uri
 
   def converted_text
-    linkify([@status_parser.title.presence, @status_parser.spoiler_text.presence, @status_parser.url || @status_parser.uri].compact.join("\n\n"))
+    title = "<h2>#{@status_parser.title}</h2>" if @status_parser.title.present?
```
Mastodon replaces `h1` to `h6` tags with `p` tags, so that is not especially useful.

Menrath replied:
I thought `<h1>` to `<h6>` tags would be replaced by `<p><strong>` tags. Thats what I found out by just comparing status, e.g. [this one](https://event-federation.eu/2023/12/20/it-is-all-about-the-community/). Maybe that processing happens already before: then of course that line should be changed accordingly.

ClearlyClaire replied:
Right, `<h2>Title</h2>` is changed to `<p><strong>Title</strong></p>`, which I guess is ok.

Menrath replied:
Thanks again a lot for reviewing! If that is the case, I would propose to revert https://github.com/mastodon/mastodon/commit/10c768c6906de7ca70ae890f99e66669eb00202 because some forks may use the `<h>` tags. Also, if in the future there are features that are affected by these changes, I think this would be more clear.

Menrath replied:
Done. As you mentioned no further aspects I'll close the conversation for now.

ClearlyClaire commented:
Thanks for your contribution. Mastodon sanitizes the HTML at a later point in the process, so you don't need to sanitize it here.

However, your PR changes how the activities are rendered, without changing the test covering that behavior. They need to be updated as well.

Menrath committed:
Adopt tests for converted objects: The convertet text contains \n\n between the converted parts.

Menrath committed:
Add test for converted object type: Event with summary and other unconverted fields.

Menrath commented:
I was not too sure, whether including the '\n\n' in the test is the desired behavior. I also added a test for a more typical object Mastodon is likely to receive.

Menrath committed:
Fix tests for converted object types test:
- newlines don't have to be escaped
- prefer single quotes strings
- fix missing space after colon

Menrath commented:
Sorry for all the force pushes, as you squash your pull requests anyway it would have been more clear without them to follow what happened.

Menrath committed:
Improvement in the comprehensibility of tests that convert an event object with a summary.

ClearlyClaire reviewed (dismissed):
I'd like a second set of eyes on this, but this looks good to me.

mjankowski commented on spec/services/activitypub/fetch_remote_status_service_spec.rb:
```diff
@@ -125,7 +125,34 @@
+
+        expect(status).to_not be_nil
+        expect(status.url).to eq 'https://foo.bar/@foo/1234'
+        expect(strip_tags(status.text)).to eq "Let's change the world\n\nWe meet on January 31st at 8pm in the FooBaar!\n\nhttps://foo.bar/@foo/1234"
```
Both this added coverage and the pre-existing are all using strip_tags in their assertions ... which I think is mostly fine for the purposes we had. That said - since this code is adding the `<h2>`-wrapped value ... should we assert for its presence (and/or any other ways it may be converted elsewhere?)

oneiros replied:
@Menrath As far as I can see there is still no test that the thing you set out to to, preserve HTML in summaries, now works.

Would you like to add one? Or would you prefer one of us doing it?

Menrath committed:
Merge branch 'main' into patch-2

Menrath committed:
Refactor text formatting logic for better modularity and reusability
as used when converting (external) object types that are converted.
- Moved title and URL formatting into separate helper methods (`formatted_title`, `formatted_url`) for improved separation of concerns.
- Enhanced code readability and maintainability by isolating formatting logic.
- Preserved original functionality while making the codebase easier to extend in the future.

Menrath committed:
Merge branch 'mastodon:main' into patch-2

Menrath committed:
ruby linting: remove double empty line

Menrath committed:
Merge branch 'mastodon:main' into patch-2

Menrath committed:
Add tests to verify processing of converted objects
- Verify that the title is rendered as <h2> for both Event and Article objects
- Ensure that the summary field supports HTML and is preserved correctly
- Ensure that the url/id of the object is present as a link

Menrath committed:
Merge branch 'main' into patch-2

mjankowski commented on spec/services/activitypub/fetch_remote_status_service_spec.rb:
```diff
@@ -125,7 +125,58 @@
+
+        expect(status).to_not be_nil
+        expect(status.url).to eq object[:id]
+        expect(status.text).to start_with "<h2>#{object[:name]}</h2>\n\n#{object[:summary]}\n\n"
```
Thanks for adding this, I think that captures what you're trying to do here.

Should the value `status.text` here always be valid/parsable html? If so, maybe add a spec helper method here to use Nokogiri to actually parse thing and then pull out out the expected values with the css helpers? Might make this slightly more readable, and would fail if we ever broke the html output.

Menrath committed:
Merge branch 'main' into patch-2

Menrath committed:
Merge branch 'main' into patch-2
